src/Features/CSharp/Portable/ValidateFormatString/CSharpValidateFormatStringDiagnosticAnalyzer.cs

```diff
@@ -4,109 +4,42 @@
-
-        private static bool TryGetArgument(
-            SemanticModel semanticModel,
-            string searchArgumentName,
-            SeparatedSyntaxList<ArgumentSyntax> arguments,
-            ImmutableArray<IParameterSymbol> parameters,
-            out ArgumentSyntax argumentSyntax)
-        {
-            argumentSyntax = null;
-
-            // First, look for a named argument that matches
-            var namedArgument = arguments.SingleOrDefault(
-                p => p.NameColon?.Name.Identifier.ValueText.Equals(searchArgumentName) == true);
-
-            if (namedArgument != null)
-            {
-                argumentSyntax = namedArgument;
-                return true;
-            }
-
-            // If no named argument exists, look for the named parameter
-            // and return the corresponding argument
-            var namedParameter = parameters.SingleOrDefault(p => p.Name.Equals(searchArgumentName) == true);
-            if (namedParameter != null)
-            {
-                // For the case string.Format("Test string"), there is only one argument
-                // but the compiler created an empty parameter array to bind to an overload
-                if (namedParameter.Ordinal >= arguments.Count)
-                {
-                    return false;
-                }
-
-                // Multiple arguments could have been converted to a single params array, 
-                // so there wouldn't be a corresponding argument
-                if (namedParameter.IsParams && parameters.Length != arguments.Count)
-                {
-                    return false;
-                }
+        protected override SyntaxKind GetInvocationExpressionSyntaxKind()
+            => SyntaxKind.InvocationExpression;
 
-                argumentSyntax = arguments[namedParameter.Ordinal];
-                return true;
-            }
+        protected override ITypeSymbol GetArgumentExpressionType(
+            SemanticModel semanticModel, 
+            SyntaxNode argsArgument)
+                => semanticModel.GetTypeInfo(((ArgumentSyntax)argsArgument).Expression).Type;
 
-            return false;
-        }
-
-        protected override bool TryGetFormatStringLiteralExpressionSyntax(
-            SemanticModel semanticModel,
-            SeparatedSyntaxList<SyntaxNode> arguments,
-            ImmutableArray<IParameterSymbol> parameters,
-            out SyntaxNode formatStringLiteralExpressionSyntax)
+        protected override SyntaxNode GetMatchingNamedArgument(
+            SeparatedSyntaxList<SyntaxNode> arguments, 
+            string searchArgumentName)
         {
-            formatStringLiteralExpressionSyntax = null;
-
-            if (!TryGetArgument(semanticModel, "format", arguments, parameters, out var formatArgumentSyntax))
+            var matchingNamedArguments = arguments.Cast<ArgumentSyntax>()
+                .Where(p => p.NameColon?.Name.Identifier.ValueText.Equals(searchArgumentName) == true);
```

Comment:
PERF: There are a lot of hidden allocations here: `Cast`, `Where`, `Lambada`, `Count()`, `Single()`.  Consider expanding to a foreach loop.

Do you explicitly need to handle the more than one case? If not, just return the first one you come across?